src/main/scala/mesosphere/marathon/upgrade/DeploymentPlanReverter.scala

```diff
@@ -55,85 +55,44 @@ private[upgrade] object DeploymentPlanReverter {
-      val removedDependencies = oldGroup.dependencies -- newGroup.dependencies
-      val addedDependencies = newGroup.dependencies -- oldGroup.dependencies
+  private[this] def revertAppChanges(
+    version: Timestamp, changes: Seq[(Option[AppDefinition], Option[AppDefinition])])(
+    g: Group): Group = {
 
-      if (removedDependencies.nonEmpty || addedDependencies.nonEmpty) {
-        if (log.isDebugEnabled)
-          log.debug(
-            s"revert dependency changes in group ${oldGroup.id}, " +
-              s"readding removed {${removedDependencies.mkString(", ")}}, " +
-              s"removing added {${addedDependencies.mkString(", ")}}")
-
-        group.copy(dependencies = group.dependencies ++ removedDependencies -- addedDependencies)
-      } else {
-        // common case, unchanged
-        group
-      }
+    changes.foldLeft(g) {
+      case (result, appUpdate) =>
+        appUpdate match {
+          case (Some(oldApp), _) => //removal or change
+            log.debug("revert to old app definition {}", oldApp.id)
+            result.updateApp(oldApp.id, _ => oldApp, version)
+          case (None, Some(newApp)) =>
+            log.debug("remove app definition {}", newApp.id)
+            result.update(newApp.id.parent, _.removeApplication(newApp.id), version)
+          case (None, None) =>
+            log.warn("processing unexpected NOOP in app changes")
+            result
+        }
     }
+  }
+}
 
-    def revertGroupAddition(result: Group, newGroup: Group): Group = {
-      // We know that all group removals for groups inside of this group have been
-      // performed before. If there are no conflicts with other deployments, we could
-      // just remove the group. Unfortunately, another deployment could have
-      // added a new sub group or sub app definition. In this case, we should not remove
-      // this group.
-
-      def normalized(group: Group): Group = group.withNormalizedVersion.withoutChildren
-      def isGroupUnchanged(group: Group): Boolean =
-        !group.containsAppsOrGroups && normalized(group) == normalized(newGroup)
-
-      result.group(newGroup.id) match {
-        case Some(unchanged) if isGroupUnchanged(unchanged) =>
-
-          log.debug("remove unchanged group {}", unchanged.id)
-          result.remove(unchanged.id, version)
-        case _ if newGroup.dependencies.nonEmpty =>
-          // group dependencies have changed
-          if (log.isDebugEnabled)
-            log.debug(s"group ${newGroup.id} has changed. " +
-              s"Removed added dependencies ${newGroup.dependencies.mkString(", ")}")
-          result.update(
-            newGroup.id,
-            group => group.copy(dependencies = group.dependencies -- newGroup.dependencies),
-            version)
-        case _ =>
-          // still contains apps/groups, so we keep it
-          result
-      }
-    }
+sealed trait DeploymentPlanGroupChangesReverter {
```

Comment:
What is the reasoning behind this trait? Could we define the methods directly in the object?